src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1121 @@
+        } else {
+            Ok(TypeName::Single(
+                self.resolve_type_for_single(value)?.to_string(),
+            ))
+        }
+    }
+
+    fn resolve_type_for_single(&self, value: &Value) -> Result<&str> {
+        let Value::Object(obj) = value else {
+            bail!("Value expected to be object");
+        };
+
+        let mut possible_types = BitVec::set_after(self.types.len());
+
+        for (field, info) in &self.fields_info {
+            if obj.contains_key(field.as_str()) {
+                *possible_types &= *info.presented_in;
+            } else {
+                *possible_types &= !*info.required_in;
+            }
+
+            if possible_types.is_empty() {
+                // no possible types, something is wrong with the resolved value
+                bail!("Failed to find corresponding type for value")
+            }
+
+            if !possible_types.is_covering_multiple_types() {
+                // we've got only one possible type so return it
+                // even despite the fact the value could be totally wrong if we check other
+                // fields. But we want to cover positive cases and we want to do
+                // it as soon as possible and the wrong value will be probably
+                // wrong anyway to use later
+                return Ok(possible_types.first_covered_type(&self.types));
+            }
+        }
+
+        // we have multiple possible types. Return the first one
+        // that is defined earlier in config
+        Ok(possible_types.first_covered_type(&self.types))
+    }
+}
+
+/// BitVecesentation for set of types if some condition is hold.
+/// The condition is BitVecesented as a bit inside the usize number
+/// where bit position from the right in binary BitVecesentation of usize
+/// is the index of type in the set and the if the value of the bit is
+/// 1 then the condition is hold
+#[derive(Copy, Clone, Default)]
+struct BitVec(usize);
+
+impl std::fmt::Debug for BitVec {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!("{:0b}", self.0))
+    }
+}
+
+impl Deref for BitVec {
+    type Target = usize;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for BitVec {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl BitVec {
+    fn new() -> Self {
+        Self(0)
+    }
+
+    fn set_all(self) -> Self {
+        Self(usize::MAX)
+    }
+
+    /// Create new BitVec where condition is hold for every type
+    fn set_after(self, id: usize) -> Self {
+        Self(self.0 | ((1 << id) - 1))
+    }
+
+    /// Create new BitVec where condition is hold
+    /// for type with passed index
+    fn set_at(self, id: usize) -> Self {
+        Self(self.0 | (1 << id))
+    }
+
+    /// Search for first type in list for which condition is hold
+    fn first_covered_type<'types>(&self, types: &'types [String]) -> &'types str {
```

Comment:
We don't need to pass the slice of String to this BitVec. Keep the BitVec generic and multi-purpose.

Reply:
We need to have better bit_vec specific names here.

Reply:
if we want to have something multi-purpose for bitvec we can use https://docs.rs/bitvec/latest/bitvec/index.html.

But here I prefer to have specific implementation for this use-case since we can built more precise and more performant solution (ofc until usize repr is enough for us, otherwise bitvec should be better)